releasenotes/notes/circuit-gates-rust-5c6ab6c58f7fd2c9.yaml

```diff
@@ -0,0 +1,79 @@
+---
+features_circuits:
+  - |
+    A native rust representation of Qiskit's standard gate library has been added. When a standard gate
+    is added to a :class:`~.QuantumCircuit` or :class:`~.DAGCircuit` it is now represented in a more
+    efficient manner directly in Rust seamlessly. Accessing that gate object from a circuit or dag will
+    return a new Python object representing the standard gate. This leads to faster and more efficient
+    transpilation and manipulation of circuits for functionality written in Rust.
+features_misc:
+  - |
+    Added a new build-time environment variable ``QISKIT_NO_CACHE_GATES`` which
+    when set to a value of ``1`` (i.e. ``QISKIT_NO_CACHE_GATES=1``) which
+    decreases the memory overhead of a :class:`.CircuitInstruction` and
+    :class:`.DAGOpNode` object at the cost of decreased runtime on multiple
+    accesses to :attr:`.CircuitInstruction.operation` and :attr:`.DAGOpNode.op`.
+    If this environment variable is set when building the Qiskit python package
+    from source the caching of the return of these attributes will be disabled.
+upgrade_circuits:
+  - |
+    The :class:`.Operation` instances of :attr:`.DAGOpNode.op`
+    being returned will not necessarily share a common reference to the
+    underlying object anymore. This was never guaranteed to be the case and
+    mutating the :attr:`~.DAGOpNode.op` directly by reference
+    was unsound and always likely to corrupt the dag's internal state tracking
+    Due to the internal refactor of the :class:`.QuantumCircuit` and
+    :class:`.DAGCircuit` to store standard gates in rust the output object from
+    :attr:`.DAGOpNode.op` will now likely be a copy instead of a shared instance. If you
+    need to mutate an element should ensure that you either do::
+
+      op = dag_node.op
+      op.params[0] = 3.14159
+      dag_node.op = op
+
+    or::
+
+      op = dag_node.op
+      op.params[0] = 3.14159
+      dag.substitute_node(dag_node, op)
+
+    instead of doing something like::
+
+      dag_node.op.params[0] = 3.14159
+
+    which will not work for any standard gates in this release. It would have
+    likely worked by chance in a previous release but was never an API guarantee.
```

Comment:
If we did this internally (we did), I'll bet others are, and it's a sufficiently "obvious" transformation that I think we need to ensure that it continues to work until there's been _at least_ a warning period.  Let's make this continue to work for 1.2 (make standard gate `params` fields a custom `list` subclass that calls back to the circuit to do the update).